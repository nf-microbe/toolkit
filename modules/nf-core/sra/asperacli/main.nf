process SRA_ASPERACLI {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'docker://carsonjm/asperacli_fastp_megahit_phables_seqkit:0.0.1' :
        'docker.io/carsonjm/asperacli_fastp_megahit_phables_seqkit:0.0.1' }"

    input:
    tuple val(meta), val(accession)

    output:
    tuple val(meta), path("${accession}*.fastq.gz") , emit: fastq
    path "versions.yml"                             , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args            = task.ext.args ?: ''
    def conda_prefix    = ['singularity', 'apptainer'].contains(workflow.containerEngine) ? "export CONDA_PREFIX=/opt/conda" : ""
    prefix              = task.ext.prefix ?: "${meta.id}"
    """
    ${conda_prefix}

    ### Download FastQ metadata
    curl -X 'GET' \\
        'https://www.ebi.ac.uk/ena/portal/api/filereport?accession=${accession}&result=read_run&fields=library_layout,fastq_md5,fastq_aspera' \\
        -H 'accept: */*' > acc_metadata.tsv

    md5=\$(awk 'NR>1 { print \$2 }' acc_metadata.tsv)
    fastq=\$(awk 'NR>1 { print \$3 }' acc_metadata.tsv)
    layout=\$(awk 'NR>1 { print \$4 }' acc_metadata.tsv)

    ### Download FastQ files
    if [[ \$layout == "PAIRED" ]];
    then
        fastq_array=(\${fastq//;/ })
        md5_array=(\${md5//;/ })

        ascp \\
            -QT -l 300m -P 33001 \\
            -i \$CONDA_PREFIX/etc/aspera/aspera_bypass_dsa.pem \\
            era-fasp@\${fastq_array[0]} \\
            ${accession}_1.fastq.gz

        echo "\${md5_array[0]}  ${accession}_1.fastq.gz" > ${prefix}_1.fastq.gz.md5
        md5sum -c ${accession}_1.fastq.gz.md5

        ascp \\
            -QT -l 300m -P 33001 \\
            -i \$CONDA_PREFIX/etc/aspera/aspera_bypass_dsa.pem \\
            era-fasp@\${fastq_array[1]} \\
            ${accession}_2.fastq.gz

        echo "\${md5_array[1]}  ${accession}_2.fastq.gz" > ${prefix}_2.fastq.gz.md5
        md5sum -c ${accession}_2.fastq.gz.md5
    else
        ascp \\
            -QT -l 300m -P 33001 \\
            -i \$CONDA_PREFIX/etc/aspera/aspera_bypass_dsa.pem \\
            era-fasp@\$fastq \\
            ${accession}.fastq.gz

        echo "\$md5  ${accession}.fastq.gz" > ${prefix}.fastq.gz.md5
        md5sum -c ${accession}.fastq.gz.md5
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ascli: 4.14.0
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    ### Download FastQ metadata
    curl -X 'GET' \\
        'https://www.ebi.ac.uk/ena/portal/api/filereport?accession=${accession}&result=read_run&fields=library_layout,fastq_md5,fastq_aspera' \\
        -H 'accept: */*' > acc_metadata.tsv

    md5=\$(awk 'NR>1 { print \$2 }' acc_metadata.tsv)
    fastq=\$(awk 'NR>1 { print \$3 }' acc_metadata.tsv)
    layout=\$(awk 'NR>1 { print \$4 }' acc_metadata.tsv)

    ### Touch empty FastQ files
    if [[ \$layout == "PAIRED" ]];
    then
        echo "" | gzip > ${accession}_1.fastq.gz
        echo "" | gzip > ${accession}_2.fastq.gz
    else
        echo "" | gzip > ${accession}.fastq.gz
    fi

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ascli: 4.14.0
    END_VERSIONS
    """
}
