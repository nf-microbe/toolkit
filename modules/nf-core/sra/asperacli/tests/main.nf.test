nextflow_process {

    name "Test Process SRA_ASPERACLI"
    script "../main.nf"
    process "SRA_ASPERACLI"

    tag "modules"
    tag "modules_nfcore"
    tag "sra"
    tag "sra/asperacli"

    test("paired end + SRR24085752") {
        when {
            process {
                """
                input[0] = [
                    [ id:'SRR24085752' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'SRR24085752' ], // meta map
                    'SRR24085752'
                ]
                """
            }
        }
        then {
            assertAll (
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}