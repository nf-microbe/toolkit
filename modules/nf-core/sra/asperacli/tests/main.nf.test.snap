{
    "paired end + SRR24085752": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752"
                        },
                        [
                            "SRR24085752_1.fastq.gz:md5,c569f3c772d91bbf8e5a27e74feb9177",
                            "SRR24085752_2.fastq.gz:md5,77913182e6a33cde6d022da8c99f3a77"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,9ccba0c53d3b31dcec82bb11e5824eb2"
                ],
                "fastq": [
                    [
                        {
                            "id": "SRR24085752"
                        },
                        [
                            "SRR24085752_1.fastq.gz:md5,c569f3c772d91bbf8e5a27e74feb9177",
                            "SRR24085752_2.fastq.gz:md5,77913182e6a33cde6d022da8c99f3a77"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9ccba0c53d3b31dcec82bb11e5824eb2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T10:30:50.152490976"
    },
    "stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "SRR24085752"
                        },
                        [
                            "SRR24085752_1.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "SRR24085752_2.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "1": [
                    "versions.yml:md5,9ccba0c53d3b31dcec82bb11e5824eb2"
                ],
                "fastq": [
                    [
                        {
                            "id": "SRR24085752"
                        },
                        [
                            "SRR24085752_1.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940",
                            "SRR24085752_2.fastq.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                        ]
                    ]
                ],
                "versions": [
                    "versions.yml:md5,9ccba0c53d3b31dcec82bb11e5824eb2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-10T10:29:43.870392593"
    }
}